Cryptolib/PublicKeyEncryption.lean

```diff
@@ -0,0 +1,70 @@
+/-
+Copyright (c) 2021 Joey Lupo. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Joey Lupo
+-/
+import Cryptolib.ToMathlib
+import Cryptolib.Uniform
+import Mathlib.Data.ZMod.Basic
+import Mathlib.Probability.Distributions.Uniform
+
+/-!
+# Security games as PMFs
+
+This file provides formal definitions for correctness and semantic security of a public
+key encryption scheme.
+-/
+
+noncomputable section
+
+/-
+G1 = public key space, G2 = private key space,
+M = message space, C = ciphertext space
+A_state = type for state A1 passes to A2
+-/
+variable {G1 G2 M C A_state: Type} [DecidableEq M]
+         (keygen : PMF (G1 × G2))
+         (encrypt : G1 → M → PMF C)
+         (decrypt : G2 → C → M)
+         (A1 : G1 → PMF (M × M × A_state))
+         (A2 : C → A_state → PMF (ZMod 2)) -- Should have G1 else how can A2 do further encryptions? Any general result about encryptions before getting challenge ciphertext...?
```

Comment:
The comment is correct; A2 needs at least G1. Depending on the attack model it might also need a decryption oracle. Without a decryption oracle this is implicitly trying to formalize IND-CPA, but it isn't documented as such.

(Some papers refer to IND-CPA as "semantic security" but that usage is at least confusing and obsolete, and arguably incorrect; see below.)